### ISSUE TITLE
New lint: Unused recursive parameter

### ISSUE THREAD

**Issue by camsteffen** [rust-lang/rust-clippy#6641]
### What it does

Checks for parameters that are passed recursively, but otherwise unused.

### Categories (optional)

- Kind: correctness (or complexity?)

### Drawbacks

None.

### Example

```rust
fn silly(a: i32, b: i32) -> i32 {
    if a > 0 {
        silly(a - 1, b)
    } else {
        0
    }
}
```

Could be written as:

```rust
fn silly(a: i32) -> i32 {
    if a > 0 {
        silly(a - 1)
    } else {
        0
    }
}
```

Make sure the parameter is passed in the same position!

Also could lint with modifications to the parameter like `silly(a - 1, b - 1)` - `b` would still be unused. But you would have to check for side effects.

Of course, it would be nice to lint on recursion involving multiple functions, but Clippy currently does not do inter-function lints.

**Comment by camsteffen** [comment]
Whoops dupe of #2624
